cpp/src/arrow/type.cc

```diff
@@ -1060,19 +1060,40 @@ struct FieldPathGetImpl {
   }
 
   template <typename T, typename GetChildren>
-  static Result<T> Get(const FieldPath* path, const std::vector<T>* children,
-                       GetChildren&& get_children, int* out_of_range_depth) {
+  static Result<T> Get(const FieldPath* path, const DataType* parent,
+                       const std::vector<T>* children, GetChildren&& get_children,
+                       int* out_of_range_depth) {
     if (path->indices().empty()) {
       return Status::Invalid("empty indices cannot be traversed");
     }
 
     int depth = 0;
-    const T* out;
+    const T* out = nullptr;
     for (int index : path->indices()) {
       if (children == nullptr) {
         return Status::NotImplemented("Get child data of non-struct array");
       }
 
+      if constexpr (std::is_same_v<T, std::shared_ptr<arrow::Field>>) {
+        auto IsBaseListType = [](const DataType& type) {
+          return dynamic_cast<const BaseListType*>(&type) != nullptr;
+        };
+
+        // For lists, we don't care about the index, jump right into the list value type.
+        // The index here is used in the kernel to grab the specific list element.
+        // Children then are fields from list type, and out will be the children from
+        // that field, thus index 0.
+        if (out != nullptr && IsBaseListType(*(*out)->type())) {
+          children = get_children(*out);
```

Comment:
I don't fully understand why we need the `get_children` here (because then we call `get_children` twice in a single iteration? since there is another call a bit below) 
Can you add a clarifying comment about that?

Reply:
(or to put it differently: why isn't this initial get_children needed when parent is a struct type? What's the difference between both?)